### ISSUE TITLE
[NODE] base64decode task failing to decode abi.encodePacked(_address) from decode_cbor task

### ISSUE THREAD

**Issue by vnavascues** [smartcontractkit/chainlink#7213]
**Description**

Hi, I'm currently testing the new `base64decode` task and it errors with:

```
failed to decode base64 string: illegal base64 data at input byte 0

input: $(decode_cbor.address)
```

I'd like to know whether I'm misusing this new task or it is a bug. Thanks

**Basic Information**

The Solidity contract sends the address base64 encoded (as a map), via:

```solidity
    function request(
        bytes32 _specId,
        uint256 _payment,
        address _address
    ) external {
        Chainlink.Request memory req = buildOperatorRequest(_specId, this.fulfill.selector);

        req.addBytes("address", abi.encodePacked(_address));

        sendOperatorRequest(req, _payment);
    }
```

The `base64decode` task input is `vEyg7adkeoq3wgYcLhGKGKk28T0=` (although after certain node version it is not visible anymore in the JSON taskRun):

```json
"taskRuns": [
    {
      "__typename": "TaskRun",
      "id": "ff223dc5-ab89-4566-b94d-40e22977986d",
      "createdAt": "2022-08-10T12:29:39.568966Z",
      "dotID": "decode_log",
      "error": null,
      "finishedAt": "2022-08-10T12:29:39.569245Z",
      "output": "{"callbackAddr":"0x8F496457cA77f2Fa7c68A6F1B2127F87c0EE453d","callbackFunctionId":"0x5508ff94","cancelExpiration":1660134868,"data":"0x676164647265737354bc4ca0eda7647a8ab7c2061c2e118a18a936f13d","dataVersion":2,"payment":100000000000000000,"requestId":"0x21b4aea459a1148203fdd41f23bb3fcd318b8c19bbb042761f6d634fd809ac46","requester":"0x8F496457cA77f2Fa7c68A6F1B2127F87c0EE453d","specId":"0x3437366462353666313238623466613862306430313538393430393662666536"}",
      "type": "ethabidecodelog"
    },
    {
      "__typename": "TaskRun",
      "id": "e744f249-c385-456f-a8ea-8fc8e6a979fa",
      "createdAt": "2022-08-10T12:29:39.569362Z",
      "dotID": "decode_cbor",
      "error": null,
      "finishedAt": "2022-08-10T12:29:39.569445Z",
      "output": "{"address":"0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d"}",
      "type": "cborparse"
    },
    {
      "__typename": "TaskRun",
      "id": "f48a0900-9b97-4e12-9a80-9e6b65dc43b6",
      "createdAt": "2022-08-10T12:29:39.569492Z",
      "dotID": "base64_decoder",
      "error": "failed to decode base64 string: illegal base64 data at input byte 0",
      "finishedAt": "2022-08-10T12:29:39.569621Z",
      "output": "null",
      "type": "base64decode"
    }
  ]
```

This is how I'd decode it using node.js (for instance in a `base64decode` external adapter):

```javascript
b64addr = "vEyg7adkeoq3wgYcLhGKGKk28T0="
`0x${Buffer.from(b64addr, "base64").toString('hex')}` // '0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d'
```

- Network: Kovan
- Node version: v1.7.0

**Environment Variables**
N/A

**Steps to Reproduce**

[Base64DecodeConsumer contract](https://kovan.etherscan.io/address/0x8F496457cA77f2Fa7c68A6F1B2127F87c0EE453d#code)
  - [Request transaction](https://kovan.etherscan.io/tx/0x55bd075e9c5c24b8152361b75b66e1d08f4bacc633f43535d43ef086485be64f)

Basic TOML spec:

```toml
type = "directrequest"
schemaVersion = 1
name = "Test base64decode task"
externalJobID = "476db56f-128b-4fa8-b0d0-15894096bfe6"
maxTaskDuration = "0s"
contractAddress = "0x878541888a928a31F9EAb4cB61DfD4e381EC2f00"
minContractPaymentLinkJuels = "10000000000000000"
observationSource = """
    decode_log          [type="ethabidecodelog"
                         abi="OracleRequest(bytes32 indexed specId, address requester, bytes32 requestId, uint256 payment, address callbackAddr, bytes4 callbackFunctionId, uint256 cancelExpiration, uint256 dataVersion, bytes data)"
                         data="$(jobRun.logData)"
                         topics="$(jobRun.logTopics)"]

    decode_cbor         [type="cborparse" data="$(decode_log.data)"]

    base64_decoder      [type="base64decode" input="$(decode_cbor.address)"]

    decode_log -> decode_cbor -> base64_decoder
"""
```

[JSON jobrun node 1.7.0](https://gist.github.com/vnavascues/7b14168dbb3d1fb02d0ad369d2a72bec)

This picture shows how the base64 string was shown on old node versions:
![image](https://user-images.githubusercontent.com/14769962/183902801-528db4c7-500e-4aa0-9ad9-542401c126c9.png)

**Additional Information**
Related issues:
https://github.com/smartcontractkit/chainlink/issues/6555
https://github.com/smartcontractkit/documentation/issues/255

**Comment by zeuslawyer** [comment]
thanks @vnavascues - will take a look.

**Comment by zeuslawyer** [comment]
@vnavascues , thanks to @aelmanaa , I am able to offer you some suggestions.  If it wasn't for his excellent help and careful explanations to me I would have struggled with this too!

1) `req.addBytes("address", abi.encodePacked(_address))`.    `.encodePacked()` returns ABI ETH spec bytes.  Bytes (binary)  is different from [base64 which is technically a binary-to-string encoding](https://en.wikipedia.org/wiki/Base64) and therefore different from ABI (bytes) encoding.   

2) Furthermore, `.encodePacked` is unpadded.  Instead we need to use `encode()` [footnote A] which pads with extra zeroes to achieve 32 bytes, and is what works on the CL Nodes [ETH ABI Decode Task](https://docs.chain.link/docs/jobs/task-types/eth-abi-decode/).    Turns out that anything that is not 32 bytes is problematic for the CBOR decoding [footnote B]. So if you use `abi.encode()` and change your TOML to use an `ethabidecode` task type instead of a `base64decode` task type, it will work. The `addBytes` method on Chainlink.Request also suggests that we can pass an address using abi.encode [footnote C]

It would look like this:
```
function request( ) external {
        Chainlink.Request memory req = buildOperatorRequest('###JOB_SPEC_ID_HERE', this.fulfill.selector);
      
        req.addBytes("address", abi.encode(address(this)));
        sendOperatorRequest(req, (1 * LINK_DIVISIBILITY) / 10);
    }
```

The relevant Decode Task for the TOML job looks like this when you abi.encode() the address.
```
abidecode [type="ethabidecode"
        abi="address address"
        data="$(decode_cbor.address)"] 
```

3) to play with the base 64 encode task, please use this [gist that @aelmanaa has produced](https://gist.github.com/aelmanaa/d09b4f94d8d67175c831390500268edf).  He has used the Open Zeppelin library to convert the address to a base64 String and send that in the request.  The relevant part of the job's TOML with the tasks looks like this:
```
observationSource = """
    decode_log          [type="ethabidecodelog"
                         abi="OracleRequest(bytes32 indexed specId, address requester, bytes32 requestId, uint256 payment, address callbackAddr, bytes4 callbackFunctionId, uint256 cancelExpiration, uint256 dataVersion, bytes data)"
                         data="$(jobRun.logData)"
                         topics="$(jobRun.logTopics)"]

    decode_cbor         [type="cborparse" data="$(decode_log.data)"]

    base64_decoder      [type="base64decode" input="$(decode_cbor.address)"]

    decode_log -> decode_cbor -> base64_decoder
```


Once again, a big thanks to @aelmanaa who helped a lot in understanding this.

[A] As per @aelmanaa's [comment in Issue #6555](https://github.com/smartcontractkit/chainlink/issues/6555#issue-1226478742),
[B]  As per @aelmanaa's [comment  in smartcontractkit/documentation#255](https://github.com/smartcontractkit/documentation/issues/255#issuecomment-1118560594)
[C] https://docs.chain.link/docs/chainlink-framework/#addbytes

**Comment by vnavascues** [comment]
@zeuslawyer @aelmanaa thank you so much for this thorough explanation and the time spent on it, amazing.

- Wrt `req.addBytes()`: I thought that it was sending the data as a base64 encoded string.
- Wrt `ethabidecode`: it is the only classic tasks I've barely used. In fact my first time was trying to move away from passing an address as `string` (this is where I found out the aelmanaa issues). Somehow, I misused `ethabidecode` (probably due to `abi.encodePacked`) and eventually found another way.

I've just successfully tested sending `address` and `address[]` via `addBytes()` and decoding them with `ethabidecode`. And I get now the usecase of `base64decode` and how send the data from the consumer contract.

Thank you!
